Add support for SomfyHeatingTemperatureInterface in Overkiz integration

egguy opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Back ported the code of somfyHeatingTemperatureInterface from ha-tahoma to HA core.
Modernised the code to use the new features of HA core



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

egguy committed:
Port of the somfyHeatingTemperatureInterface

Ported the code from ha-tahoma to HA core. Modernized the code to use the new features

nyroDev commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+        )
+
+        if action := OverkizCommandParam(current_operation):
+            return action
```
You shouldn't retrurn an OverkizCommandParam directly.  
I think you should use `OVERKIZ_TO_HVAC_ACTION` here

nyroDev commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set new temperature."""
+        temperature = kwargs.get(ATTR_TEMPERATURE)
```
Please use `kwargs[ATTR_TEMPERATURE]`

nyroDev reviewed (changes requested):
Thanks for this PR, it looks pretty good.

egguy committed:
Use the mapping dict when returning hvac action.

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+        if mode := OVERKIZ_TO_HVAC_MODES[OverkizCommandParam(state)]:
+            return mode
+
+        if state is not None:
+            # Unknown and potentially a new state, log to make it easier to report
+            _LOGGER.warning(
+                "Overkiz %s state unknown: %s",
+                OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_ACTIVE_MODE,
+                state,
+            )
```
Do we need this in core? I would say that we now know which states are there; and otherwise it will be in the log.

Quentame replied:
If we do add it, we must also in other core Overkiz entities, which do not have that at the moment

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+                OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_SETPOINT_MODE
+            ),
+        )
+        return OVERKIZ_TO_PRESET_MODES[OverkizCommandParam(state)]
```
I would prefer to not wrap everything with OverkizCommandParam, we don't do this in other implementations..

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+            OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_SETPOINT_MODE
+        )
+
+        if mode not in MODE_COMMAND_MAPPING:
+            _LOGGER.error("Unknown temperature mode: %s", mode)
+            return None
```
```suggestion
```

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+            return None
+
+        return await self.executor.async_execute_command(
+            MODE_COMMAND_MAPPING[OverkizCommandParam(mode)], temperature
```
Same

tetienne commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+    @property
+    def hvac_mode(self) -> str:
+        """Return hvac operation i.e. heat, cool mode."""
+        if (
```
You can now avoid the cast with something like:

`self.device.states[OverkizState.CORE_ON_OFF].value_as_str`

tetienne commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+    @property
+    def target_temperature(self) -> float | None:
+        """Return the target temperature."""
+        if self.preset_mode not in PRESET_MODES_TO_OVERKIZ:
```
I don’t think this case can happen.

tetienne commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+        if mode not in MAP_PRESET_TEMPERATURES:
+            return None
+
+        return cast(float, self.executor.select_state(MAP_PRESET_TEMPERATURES[mode]))
```
you can do: `self.device.states[MAP_PRESET_TEMPERATURES].values_as_float`

Quentame commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,203 @@
+    )
+    _attr_hvac_modes = [*HVAC_MODES_TO_OVERKIZ]
+    _attr_preset_modes = [*PRESET_MODES_TO_OVERKIZ]
+    # Both min and max temp values have been retrieved from the Somfy Application.
+    _attr_min_temp = 15.0
+    _attr_max_temp = 26.0
```
And therefore can't be get programmatically ?

iMicknl replied:
I assume not, we tried to gather this from attributes where available, but many devices don't have these attributes.

@egguy do you have your diagnostics file? We can have a look.

iMicknl replied:
Have checked the diagnostics file, unfortunately these values do not exists programmatically.

egguy committed:
Cleaned up the code.
- Removed casts
- Don't use OverkizCommandParam directly

egguy committed:
Keep the alpha ordering in the climate_entities

Co-authored-by: Quentame <polletquentin74@me.com>

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,183 @@
+from ..coordinator import OverkizDataUpdateCoordinator
+from ..entity import OverkizEntity
+
+_LOGGER = logging.getLogger(__name__)
```
Please import `LOGGER` from `.const`, no need to declare again.

iMicknl commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,183 @@
+    PRESET_AWAY: OverkizState.CORE_SECURED_POSITION_TEMPERATURE,
+}
+
+MODE_COMMAND_MAPPING: dict[str, str] = {
```
```suggestion
SETPOINT_MODE_TO_OVERKIZ_COMMAND: dict[str, str] = {
```
nit: perhaps a bit more descriptive name?

iMicknl reviewed (changes requested):
Thanks a lot for your contribution @egguy! Would you be able to make some small changes as mentioned by the others + my review? Would be great to have this one in core as well.

egguy commented:
Hello, sorry for the response delay, I changed country.

I will try to make the changes as soon as possible. 

As for the diag file, I will try. The box has been shutdown for 2 months :s

iMicknl commented:
@egguy great! Let us know if we can help getting this PR in :)

egguy committed:
Add the suggested changes

- Use the logger from the integration
- Rename MODE_COMMAND_MAPPING to SETPOINT_MODE_TO_OVERKIZ_COMMAND
- Update the code to use the enum for the temperature

iMicknl commented:
> There hasn't been any activity on this pull request recently. This pull request has been automatically marked as stale because of that and will be closed if no further activity occurs within 7 days. Thank you for your contributions.

Not stale. @egguy would this one be ready for review, I see you addressed all current feedback?

egguy commented:
I think this should be ready for a review. Is there any change request I have missed?

iMicknl reviewed (approved):
LGTM! (and understood you own this device yourself + have tested all functionality)

Quentame commented on homeassistant/components/overkiz/const.py:
```diff
@@ -99,6 +99,7 @@
     UIWidget.RTS_GENERIC: Platform.COVER,  # widgetName, uiClass is Generic (not supported)
     UIWidget.SIREN_STATUS: None,  # widgetName, uiClass is Siren (siren)
     UIWidget.SOMFY_THERMOSTAT: Platform.CLIMATE,  # widgetName, uiClass is HeatingSystem (not supported)
+    UIWidget.SOMFY_HEATING_TEMPERATURE_INTERFACE: Platform.CLIMATE,  # widgetName, uiClass is HeatingSystem (not supported)
```
```suggestion
    UIWidget.SOMFY_HEATING_TEMPERATURE_INTERFACE: Platform.CLIMATE,  # widgetName, uiClass is HeatingSystem (not supported)
    UIWidget.SOMFY_THERMOSTAT: Platform.CLIMATE,  # widgetName, uiClass is HeatingSystem (not supported)
```

Quentame commented on homeassistant/components/overkiz/climate_entities/somfy_heating_temperature_interface.py:
```diff
@@ -0,0 +1,179 @@
+    OverkizCommandParam.SECURED: PRESET_AWAY,
+    OverkizCommandParam.ECO: PRESET_ECO,
+    OverkizCommandParam.COMFORT: PRESET_COMFORT,
+    OverkizCommandParam.FREE: PRESET_NONE,
```
```suggestion
    OverkizCommandParam.FREE: PRESET_NONE,
    OverkizCommandParam.FROSTPROTECTION: PRESET_FROST_PROTECTION,
```
It seems this device works with frost protection, as https://github.com/home-assistant/core/blob/dev/homeassistant/components/overkiz/climate_entities/atlantic_electrical_heater.py or https://github.com/home-assistant/core/blob/dev/homeassistant/components/overkiz/climate_entities/atlantic_pass_apc_heating_zone.py or https://github.com/home-assistant/core/blob/dev/homeassistant/components/overkiz/climate_entities/valve_heating_temperature_interface.py

Quentame replied:
That can be done in a new PR as it needs tests.

Quentame committed:
Update homeassistant/components/overkiz/const.py

Quentame reviewed (changes requested):
`mypy` pipeline job needs to be fixed

egguy committed:
Updated the typing of the functions + dicts
